### PR TITLE
fix(asgi): add `finally` statement for stream.close()

### LINKED ISSUE #1943: [ASGI] close() not called if iterating over resp.stream raises an exception
At the time of writing, the `close()` method of [`resp.stream`](https://falcon.readthedocs.io/en/stable/api/request_and_response_asgi.html#falcon.asgi.Response.stream) is only called if the provided generator is successfully exhausted (or yields `None`).

However, the primary use case of the `close()` method is freeing up resources, so one might expect the method to be called regardless. It is also what [PEP 3333 compliant](https://www.python.org/dev/peps/pep-3333/#specification-details) WSGI servers do.

I would suggest introducing a `finally` clause, and moving the invocation of `close()` there.

Make sure to also write proper regression tests:
>  Kurt Griffiths @kgriffs
> We also need a test that forces stream to raise an error.
> (and ensures close() is called)

### PR DESCRIPTION
**Pull request by dimucciojonathan**
# Summary of Changes

This PR has the same implementation from the closed PR #1948, except this is on a new branch.

# Related Issues

Closes #1943

### CODE REVIEW

File: falcon/asgi/app.py [hunk 1]
```diff
@@ -698,9 +698,9 @@ async def watch_disconnect():
                         '__aiter__ method. Error raised while iterating over '
                         'Response.stream: ' + str(ex)
                     )
-
-            if hasattr(stream, 'close'):
-                await stream.close()
+                finally:
+                    if hasattr(stream, 'close'):
```
Comment (vytas7) [inline falcon/asgi/app.py:702]: We shouldn't move `.close()` in here. This way, `.close()` is only attempted for the second branch of the main conditional statement (whether `stream` has `read()` or not). That's why the tests are failing too.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Hi again, that's a good start.

However, as pointed out inline, you should keep the original scope of `.close()`, attempting it for file-like objects with a `.read()` method too (that's probably even the primary use case for `.close()`).

- [ ] Maintain the original `.close()` scope
- [ ] Add unit tests asserting that `.close()` is attempted even upon exception while iterating over `stream`, also when calling its `read()` method.
- [x] Add a news fragment describing the bug fix

File: falcon/asgi/app.py (hunk 1 above)
Comment (dimucciojonathan) [reply falcon/asgi/app.py:702]: Thanks, this makes a lot of sense!

**Comment by dimucciojonathan** [comment]
@vytas7 

I have another beginner question / progress for #1943. When adding tests for `.close()` when an exception occurs, it seems like the best approach is adding assertions to the `test_genfunc_error` and `test_nongenfunc_error` functions in `test_hello_asgi.py` since this is where the exception is tested. 

I see that `.close()` is tested directly under these functions, but a different `resource` variable is used. So if I'm on the right path here, how can I verify the `HelloResource` is closed?

I will also update the branch containing the PR with the change to complete bulletpoint 1 of your previous comment.

**Review by prabin-lp — DISMISSED** [review]

**Comment by vytas7** [comment]
Hi again @dimucciojonathan! Just checking if you are planning to continue working on this issue, or could we possibly join forces with another contributor?

**Comment by dimucciojonathan** [comment]
Hi @vytas7 , thank you for asking. I am all out of ideas for how to properly close this issue, so feel free to assign another contributor and I will follow along!

**Comment by vytas7** [comment]
Thanks for the prototyping so far @dimucciojonathan -- maybe this issue is not that trivial as it looks at first glance.
For instance, in addition to what has been prototyped above, one non-obvious choice is whether to call `await close()` as it is now, or to schedule that later asynchronously.

@wsz you had some ideas how to resolve this, would you like to take over this?

**Comment by vytas7** [comment]
Hi again @dimucciojonathan, I believe this is how one of the ways to fix this could look like.
Still remaining are a Towncrier newsfragment, and proper unit test coverage for the change.

**Review by vytas7 — APPROVED** [review]
I've taken the liberty of adding further patches to this PR; hopefully it's ready for review now.

**Review by CaselIT — APPROVED** [review]
lgtm

**Pull request merged**